File: OOP/Coursework/CPPConsoleApp/CPPConsoleApp/BigIntArray.cpp
#include "stdafx.h"
#include "BigIntArray.h"
#include <iomanip>

BigIntArray::BigIntArray(void)
{
	pBigInts = new BigInt[0];
	Count = 0;
	uiPosition = 0;
}

BigIntArray::BigIntArray(int nCount)
{
	uiPosition = 0;
	Count = nCount;
	pBigInts = new BigInt[nCount];
}

BigIntArray::~BigIntArray(void)
{
	if (pBigInts) delete[] pBigInts;
}

void BigIntArray::Input()
{
	for (int i = 0; i < Count; i++)
	{
		cout << "×èñëî #" << i + 1 << endl;
		cin >> *(pBigInts + i);
	}

}

BigInt* BigIntArray::GetBigInts()
{
	return pBigInts;
}

int BigIntArray::GetCount()
{
	return Count;
}

void BigIntArray::Show()
{
	cout << endl;
	cout << "Âìiñò:" << endl;
	for (int i = 0; i < Count; i++)
		cout << *(pBigInts + i) << " ";
}

BigInt BigIntArray::Sum()
{
	BigInt sum(0);
	
	for (int i = 0; i < Count; i++)
	{
		sum = sum + pBigInts[i];
	}
	return sum;
}

BigInt* BigIntArray::begin()
{
	if (pBigInts == NULL || Count <= 0)
	{
		throw runtime_error("Bad pointer");
	}

	uiPosition = 0;

	return pBigInts;
}

BigInt* BigIntArray::end()
{
	if (pBigInts == NULL || Count <= 0)
	{
		throw runtime_error("Bad pointer");
	}

	return pBigInts + Count;
}

File: _OLD-21/OOP/Coursework/CSharpApp/CSharpApp/BigIntArray.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Collections;

namespace BigIntNumbersCalculator
{
    public class BigIntArray : IEnumerable, IEnumerator
    {
        private List<BigInt> BigInts;
        private int Count = 0;
        private int index = -1;

        public BigIntArray() //Конструктор без параметрів
        {
            BigInts = new List<BigInt>();
        }
        public BigIntArray(int nCount) //Конструктор з параметром
        {
            BigInts = new List<BigInt>();
            Count = nCount;
        }
        // Реалізуємо інтерфйес IEnumerable
        public IEnumerator GetEnumerator()
        {
            foreach (BigInt number in BigInts)
            {
                yield return number;
            }
        }

        // Реалізуємо інтерфйес IEnumerator
        public bool MoveNext()
        {
            if (index == BigInts.Count() - 1)
            {
                Reset();
                return false;
            }

            index++;
            return true;
        }

        public void Reset()
        {
            index = -1;
        }

        public object Current
        {
            get
            {
                return BigInts.ElementAt(index);
            }
        }

        public void Input() //Метод для вводу даних
        {
            for (int i = 0; i < Count; i++)
            {
                Console.WriteLine("Запис #{0}", i + 1);
                BigInt number = new BigInt();
                //BigInt.Input();
                BigInts.Add(number);
            }
        }
        void Show() //Метод для виведення даних
        {
        }
        ~BigIntArray()  //Деструктор
        { }
        public List<BigInt> GetBigInts() //Повертає вказівник на елемент масиву
        {
            return BigInts;
        }

        // Public methods.
        public void AddBigIntNumber(string _number)
        {
            BigInts.Add(new BigInt(_number, 10));
        }

        public int GetCount() // Метод для читання поля  - кількість елементів масиву
        {
            return BigInts.Count;
        }
        public BigInt Sum()
        {
            BigInt sum = new BigInt();
            for (int i = 0; i < BigInts.Count; i++)
               sum = sum + BigInts[i];
            return sum;
        }

    }
}
